### ISSUE TITLE
gemma4: service level nondeterminism cannot be told apart from a data bug

### ISSUE THREAD

**Issue by FeathBow** [xiaguan/pegainfer#900]
## Description

Submitting the same eight requests twice returns the same tokens on a quiet machine and sometimes does not on a busy one. Two causes produce that signature. The benign one is the schedule: which rows share a step moves with load, the kernels pick shapes and reduction orders by batch width, and per-step differences accumulate through the KV cache. The serious one is a row reading metadata that is not its own: the page tables and the position array are indexed by a row's place in the batch, and the engine moves that place whenever a request retires out of the middle.

The serving path cannot tell the two apart, because the engine decides when a request joins a batch and that decision moves with the machine's load. A live determinism check is therefore only meaningful on an idle device, which is exactly when nobody is looking.

## What was observed

Single GPU (sm_89, x86_64), CUDA 12.9, 12B checkpoint, greedy (`temperature 0`). A client submits the same eight prompts together twice and compares each row's tokens and per-token logprobs.

On an otherwise idle card the check holds: one batch repeated three times comes back bit-identical, max per-token |dlogprob| 0.000000, and repeated runs of the batch-twice check all pass 8/8.

With an unrelated training job holding the card at 50 to 60 percent utilization, 10 rounds of the same check give **4/10 fully identical**, and the divergences have two distinct shapes:

```
row 6: first differs at token 13, top-2 gap there 0.1250 nat (' meters' vs ' (')
row 6: first differs at token 13, top-2 gap there 0.0000 nat (' (' vs ' meters')
row 2: first differs at token 9,  top-2 gap there 2.3125 nat ('\n' vs ' a')
```

The token-13 flips are near-ties and unremarkable. The token-9 flip is not: the winner led by 2.3 nats, far beyond what a tie-break explains, which is what forced this question. A captured instance of the same position shows the scale — decoding alone, the row ranks `' a'` top-1 at logprob −0.22 with a 1.94 nat lead; in a batch of eight it emits `'\n'` and `' a'` is not in the returned top five.

Two scoping facts narrow where the cause can live. No divergence ever appeared at the first decoded token, which comes out of a prefill; every one sat at decode position 9 or later, so whatever moves accumulates across batched decode steps. And holding the batch width trajectory fixed while replacing the other rows' content and lengths leaves a row bit-identical (the table in `docs/models/gemma4/serving.md`), so the companions' data does not reach the row — but that table was measured at the API, where the trajectory can only be held approximately and only on an idle machine.

## Proposed Plan

- Drive `decode_batch_step` directly with the composition pinned, in the one shape that can hide an indexing error: a ragged batch, with rows at unrelated frontiers and page counts and one of them past the sliding window with its front released. A uniform batch cannot ask this — rows with the same prompt hold the same metadata and the same content, so even a read of a neighbour's page returns the right bytes.
- Run it three times: the same order twice, so any nondeterminism in the step itself shows, and reversed once, because a row's place in the batch is what the page tables, positions and origins are keyed by and what the engine changes when a request retires out of the middle. Every request must get bit-identical logits all three times.

## Acceptance Criteria

- The probes decide the question on a machine with unrelated load, where the live check cannot.
